internal/pkg/chainsync/internal/syncer/syncer.go

```diff
@@ -289,6 +299,15 @@ func (syncer *Syncer) syncOne(ctx context.Context, grandParent, parent, next blo
 		return err
 	}
 
+	// Now that the tipset is validated preconditions are satisfied to check
+	// consensus faults
+	for i := 0; i < next.Len(); i++ {
+		err := syncer.faultDetector.CheckBlock(next.At(i), parent)
+		if err != nil {
+			return err
+		}
+	}
+
```

Comment:
Since this method proceeds whether or not a fault is detected, I would move this to the end, after storing all state. It's kind of an aside, could even be async.